lib/pos_blockchain.go

```diff
@@ -0,0 +1,1134 @@
+		return err
+	}
+	return nil
+}
+
+// isBlockTimestampValidRelativeToParentPoS validates that the block's timestamp is
+// greater than its parent's timestamp.
+func (bc *Blockchain) isBlockTimestampValidRelativeToParentPoS(block *MsgDeSoBlock) error {
+	// Validate that the timestamp is not less than its parent.
+	parentBlock, exists := bc.blockIndexByHash[*block.Header.PrevBlockHash]
+	if !exists {
+		// Note: this should never happen as we only call this function after
+		// we've validated that all ancestors exist in the block index.
+		return RuleErrorMissingParentBlock
+	}
+	if block.Header.TstampNanoSecs < parentBlock.Header.TstampNanoSecs {
+		return RuleErrorPoSBlockTstampNanoSecsTooOld
+	}
+	return nil
+}
+
+// isProperlyFormedBlockPoS validates the block at a surface level. It checks
+// that the timestamp is valid, that the version of the header is valid,
+// and other general integrity checks (such as not malformed).
+func (bc *Blockchain) isProperlyFormedBlockPoS(block *MsgDeSoBlock) error {
+	// First make sure we have a non-nil header
+	if block.Header == nil {
+		return RuleErrorNilBlockHeader
+	}
+
+	// Make sure we have a prevBlockHash
+	if block.Header.PrevBlockHash == nil {
+		return RuleErrorNilPrevBlockHash
+	}
+
+	// Timestamp validation
+	// First make sure we have a non-nil header
+	if block.Header == nil {
+		return RuleErrorNilBlockHeader
+	}
+
+	// Make sure we have a prevBlockHash
+	if block.Header.PrevBlockHash == nil {
+		return RuleErrorNilPrevBlockHash
+	}
+
+	// Timestamp validation
+	// TODO: Add support for putting the drift into global params.
+	if block.Header.TstampNanoSecs > uint64(time.Now().UnixNano())+bc.params.DefaultBlockTimestampDriftNanoSecs {
+		return RuleErrorPoSBlockTstampNanoSecsInFuture
+	}
+
+	// Header validation
+	if block.Header.Version != HeaderVersion2 {
+		return RuleErrorInvalidPoSBlockHeaderVersion
+	}
+
+	// Malformed block checks
+	// All blocks must have at least one txn
+	if len(block.Txns) == 0 {
+		return RuleErrorBlockWithNoTxns
+	}
+	// Must have non-nil TxnConnectStatusByIndex
+	if block.TxnConnectStatusByIndex == nil {
+		return RuleErrorNilTxnConnectStatusByIndex
+	}
+
+	// Must have TxnConnectStatusByIndexHash
+	if block.Header.TxnConnectStatusByIndexHash == nil {
+		return RuleErrorNilTxnConnectStatusByIndexHash
+	}
+
+	// Make sure the TxnConnectStatusByIndex matches the TxnConnectStatusByIndexHash
+	if !(HashBitset(block.TxnConnectStatusByIndex).IsEqual(block.Header.TxnConnectStatusByIndexHash)) {
+		return RuleErrorTxnConnectStatusByIndexHashMismatch
+	}
+
+	// Require header to have either vote or timeout QC
+	isTimeoutQCEmpty := block.Header.ValidatorsTimeoutAggregateQC.isEmpty()
+	isVoteQCEmpty := block.Header.ValidatorsVoteQC.isEmpty()
+	if isTimeoutQCEmpty && isVoteQCEmpty {
+		return RuleErrorNoTimeoutOrVoteQC
+	}
+
+	if !isTimeoutQCEmpty && !isVoteQCEmpty {
+		return RuleErrorBothTimeoutAndVoteQC
+	}
+
+	if len(block.Txns) == 0 {
```

Comment:
Looks like you already checked this above.

Reply:
yep - will delete.